README.md
# CodeGPT: AI-Assisted Git Commit Message Generator

This VS Code extension use the [CodeGPT](https://github.com/appleboy/CodeGPT) CLI tool that generate git commit messages for you using ChatGPT AI (`gpt-3.5-turbo`, `gpt-4` model) right in the SCM inputBox.

## Install

1. Install [Visual Studio Code Insiders](https://code.visualstudio.com/insiders/)

    It because this extension used the `contribSourceControlInputBoxMenu` [Proposed API](https://code.visualstudio.com/api/advanced-topics/using-proposed-api). So you can only install and use the extension in VS Code 1.85+ which is Insider version for now.

    After the `contribSourceControlInputBoxMenu` proposed API been released, the extension will publish to the VS Code Marketplace.

2. Download [codegpt-0.1.0.vsix](https://github.com/doggy8088/vscode-codegpt/releases/download/0.1.0/codegpt-0.1.0.vsix)

3. Use `Extensions: Install from VSIX` command to install the VSIX.

    ![Extensions: Install from VSIX](images/install-from-vsix.jpg)

4. Start VS Code Insider by passing `--enable-proposed-api doggy8088.codegpt` to the arguments.

    ```sh
    code-insiders --enable-proposed-api doggy8088.codegpt
    ```

## Usage

You can simply hit the button beside the Message field of the SCM pane in the Visual Studio Code.

![Generate Commit Message with CodeGPT](images/screenshot-1.jpg)

Then then `codegpt commit --preview` will run and filled the generated message in the input box.

## Limitation

This extension is highly dependant with [CodeGPT](https://github.com/appleboy/CodeGPT) CLI tool. You have to install [CodeGPT](https://github.com/appleboy/CodeGPT) and fully configured first. Please make sure `codegpt` executable path is in your `PATH` environment variable.

## Options

TODO

## Links

- Visual Studio Code Extension API
  - [VS Code API](https://code.visualstudio.com/api/references/vscode-api)
  - [Source Control API](https://code.visualstudio.com/api/extension-guides/scm-provider#scm-input-box)
  - [Using Proposed API](https://code.visualstudio.com/api/advanced-topics/using-proposed-api)
    - [vscode.proposed.contribSourceControlInputBoxMenu.d.ts](https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.contribSourceControlInputBoxMenu.d.ts)
- StackOverflow
  - [vscode extensions - How do I create a custom button in the git Message field of the VS Code Source Control tab? - Stack Overflow](https://stackoverflow.com/a/77595771/910074)
- GitHub
  - [SCM: When multiple `scm/inputBox` been registered, the button shifted · Issue #199921 · microsoft/vscode](https://github.com/microsoft/vscode/issues/199921)
  - [SourceControl - SourceControlInputBoxValueProvider API proposal · Issue #195474 · microsoft/vscode](https://github.com/microsoft/vscode/issues/195474)
  - [SourceControl - `scm/inputBox` menu proposal (#195475) · Alex0007/vscode@6e87542](https://github.com/Alex0007/vscode/commit/6e87542ae69913b011f13f980683979fa7c94a04)

CHANGELOG.md
# Change Log

All notable changes to the **CodeGPT: AI-Assisted Git Commit Message Generator** will be documented in this file.

## 0.1.0 - 2023-12-04

* Initial release


src/extension.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import { exec } from 'child_process';
import { fileURLToPath } from 'url';

let logger: vscode.OutputChannel;

export function activate(ctx: vscode.ExtensionContext) {
  logger = vscode.window.createOutputChannel('CodeGPT');
  logger.appendLine('CodeGPT 已啟動');

  ctx.subscriptions.push(
    // https://stackoverflow.com/a/77595771/910074
    vscode.commands.registerCommand('codegpt.generateCommitMessage', async (rootUri: vscode.Uri, context: any, s) => {

      // vscode extensions - How to access the api for git in visual studio code - Stack Overflow
      // https://stackoverflow.com/a/52423161/910074
      const gitExtension = vscode.extensions.getExtension('vscode.git')?.exports;
      const api = gitExtension.getAPI(1);
      const repo = api.repositories[0];
      logger.appendLine(`Git Repo Root ${repo.rootUri}`);

      // repo: repository,rootUri,inputBox,state,ui
      // logger.appendLine(`repo: ${Object.keys(repo)}`);

      const workingDir = fileURLToPath(repo.rootUri.toString());
      process.chdir(workingDir);

      logger.appendLine(`Working directory: ${process.cwd()}`);

      const commandBuilder = (command: string) => {
        return {
          withOption: (option: string, value: string) => {
            command += ` ${option} ${value}`;
            return commandBuilder(command);
          },
          build: () => {
            return command;
          }
        };
      };

      // TODO: 這裡應該要用 vscode.workspace.getConfiguration('codegpt').get('codegptPath') 來取得執行檔路徑
      const commitCommand = commandBuilder('codegpt commit')
        .withOption('--preview', '')
        // TODO: 這裡應該要用 vscode.workspace.getConfiguration('codegpt').get('diffUnified') 來取得 unified 參數
        .withOption('--diff_unified', '5')
        .build();

      exec(commitCommand, (error, stdout, stderr) => {
        if (error) {
          logger.appendLine(`Error: ${error.message}`);
          vscode.window.showErrorMessage(error.message);
          return;
        }

        if (stderr) {
          logger.appendLine(`stderr: ${stderr}`);
          vscode.window.showErrorMessage(stderr);
          return;
        }

        logger.appendLine(`stdout: ${stdout}`);

        getCommitMessage().then((commitMessage) => {
          repo.inputBox.value = commitMessage;
        });
      });

    })
  );
}

function getCommitMessage(): Promise<string> {
  return new Promise((resolve, reject) => {
    const commitEditMsgPath = '.git/COMMIT_EDITMSG';
    fs.readFile(commitEditMsgPath, 'utf8', (err, data) => {
      if (err) {
        logger.appendLine(`Error reading ${commitEditMsgPath}: ${err}`);
        vscode.window.showErrorMessage(`Error reading ${commitEditMsgPath}: ${err}`);
        reject(err);
        return;
      }
      const commitMessage = data.trim();
      logger.appendLine(`Commit message: ${commitMessage}`);
      resolve(commitMessage);
    });
  });
}

export function deactivate() { }
